src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/MetadataService.kt

```diff
@@ -0,0 +1,189 @@
+import com.amazon.opendistroforelasticsearch.indexmanagement.indexstatemanagement.util.revertISMMetadataID
+import com.amazon.opendistroforelasticsearch.indexmanagement.util.OpenForTesting
+import org.apache.logging.log4j.LogManager
+import org.elasticsearch.ExceptionsHelper
+import org.elasticsearch.action.DocWriteRequest
+import org.elasticsearch.action.bulk.BackoffPolicy
+import org.elasticsearch.action.bulk.BulkItemResponse
+import org.elasticsearch.action.bulk.BulkRequest
+import org.elasticsearch.action.bulk.BulkResponse
+import org.elasticsearch.action.support.master.AcknowledgedResponse
+import org.elasticsearch.client.Client
+import org.elasticsearch.cluster.block.ClusterBlockException
+import org.elasticsearch.cluster.service.ClusterService
+import org.elasticsearch.common.unit.TimeValue
+import org.elasticsearch.index.Index
+import org.elasticsearch.rest.RestStatus
+import java.lang.Exception
+
+/**
+ * When all nodes have same version IM plugin, MetadataService starts to move
+ * metadata from cluster state to metadata index
+ */
+@OpenForTesting
+class MetadataService(
+    private val client: Client,
+    private val clusterService: ClusterService,
+    private val skipExecution: SkipExecution,
+    private val imIndices: IndexManagementIndices
+) {
+    private val logger = LogManager.getLogger(javaClass)
+
+    @Volatile private var runningLock = false // in case 2 moveMetadata() process running
+
+    private val successIndexedIndices = mutableSetOf<metadataDocID>()
+    private var failedToIndexIndices = mutableMapOf<metadataDocID, BulkItemResponse.Failure>()
+    private var failedToCleanIndices = mutableListOf<Index>()
+
+    private var counter = 0
+    @Volatile private var finishFlag = false
+    // used in coordinator sweep to cancel scheduled process
+    fun isFinished() = finishFlag
+
+    @Volatile
+    private var retryPolicy =
+        BackoffPolicy.constantBackoff(TimeValue.timeValueMillis(50), 3)
+
+    // just implement the business logic here, no need to do those listen to settings
+    // schedule a runnable
+    suspend fun moveMetadata() {
+        if (runningLock) {
+            logger.info("There is a move metadata process running...")
+            return
+        }
+        runningLock = true
+
+        if (skipExecution.flag) {
+            logger.info("Cluster still has nodes running old version ISM plugin, skip execution on new nodes until all nodes upgraded")
+            runningLock = false
+            return
+        }
+
+        val indicesMetadata = clusterService.state().metadata.indices
+        val clusterStateMetadata = indicesMetadata.map {
+            it.key to it.value.getManagedIndexMetadata()
+        }.filter { it.second != null }.distinct().toMap()
+        // filter out previous failedToClean indices which already been indexed
+        clusterStateMetadata.filter { it.key !in failedToCleanIndices.map { index -> index.name } }
```

Comment:
Doesn't filter return a new map here? Doesn't seem like it's being used

Reply:
miss that, fixed